File: core/src/main/java/com/gmail/marcosav2010/module/ModuleDescriptor.java
package com.gmail.marcosav2010.module;

import com.gmail.marcosav2010.communicator.packet.handling.listener.PacketListener;
import org.atteo.classindex.IndexAnnotated;

import java.lang.annotation.*;

@Documented
@Retention(value = RetentionPolicy.RUNTIME)
@Target(value = ElementType.TYPE)
@IndexAnnotated
public @interface ModuleDescriptor {

    boolean load() default true;

    String name();

    int priority() default 0;

    Class<? extends ModuleScope> scope();

    Class<? extends PacketListener>[] listeners();
}

File: core/src/main/java/com/gmail/marcosav2010/module/ModuleDescriptor.kt
package com.gmail.marcosav2010.module

import com.gmail.marcosav2010.communicator.packet.handling.listener.PacketListener
import org.atteo.classindex.IndexAnnotated
import kotlin.reflect.KClass

@MustBeDocumented
@kotlin.annotation.Retention(AnnotationRetention.RUNTIME)
@Target(AnnotationTarget.ANNOTATION_CLASS, AnnotationTarget.CLASS)
@IndexAnnotated
annotation class ModuleDescriptor(val load: Boolean = true,
                                  val name: String,
                                  val priority: Int = 0,
                                  val scope: KClass<out ModuleScope>,
                                  val listeners: Array<KClass<out PacketListener>>)